src/helpers/format/format_parsing.ts

```diff
@@ -0,0 +1,315 @@
+  const internalFormat =
+    tokens && areValidDateFormatTokens(tokens) ? { type: "date", tokens } : undefined;
+  if (!internalFormat) {
+    return undefined;
+  }
+  if (
+    internalFormat.tokens.length &&
+    internalFormat.tokens.every((token) => token.type === "DATE_PART" && token.value === "a")
+  ) {
+    throw new Error("Invalid date format");
+  }
+  const convertedTokens = convertTokensToMinutesInDateFormat(internalFormat.tokens);
+  return { type: "date", tokens: convertedTokens };
+}
+
+function tokensToTextInternalFormat(
+  tokens: FormatToken[] | undefined
+): TextInternalFormat | undefined {
+  return tokens && areValidTextFormatTokens(tokens) ? { type: "text", tokens } : undefined;
+}
+
+/**
+ * Replace in place tokens "mm" and "m" that denote minutes in date format with "MM" to avoid confusion with months.
+ *
+ * As per OpenXML specification, in date formats if a data token "m" or "mm" is followed by a date token "s" or
+ * preceded by a data token "h", then it's not a month but an minute.
+ */
+function convertTokensToMinutesInDateFormat(tokens: DateInternalFormat["tokens"]) {
+  const dateParts = tokens.filter((token) => token.type === "DATE_PART") as DatePartToken[];
+  for (let i = 0; i < dateParts.length; i++) {
+    if (!dateParts[i].value.startsWith("m") || dateParts[i].value.length > 2) {
+      continue;
+    }
+    if (dateParts[i - 1]?.value.startsWith("h") || dateParts[i + 1]?.value.startsWith("s")) {
+      dateParts[i].value = dateParts[i].value.replaceAll("m", "M");
+    }
+  }
+  return tokens;
+}
+
+export function convertInternalFormatToFormat(internalFormat: MultiPartInternalFormat): Format {
+  return [
+    internalFormatPartToFormat(internalFormat.positive),
+    internalFormatPartToFormat(internalFormat.negative),
+    internalFormatPartToFormat(internalFormat.zero),
+    internalFormatPartToFormat(internalFormat.text),
+  ]
+    .filter(isDefined)
+    .join(";");
+}
+
+function internalFormatPartToFormat(
+  internalFormat: InternalFormat | undefined
+): Format | undefined {
+  if (!internalFormat) {
+    return undefined;
+  }
+  let format = "";
+  const tokens =
+    internalFormat.type !== "number"
+      ? internalFormat.tokens
+      : numberInternalFormatToTokenList(internalFormat);
+  for (let token of tokens) {
+    switch (token.type) {
+      case "ESCAPED_STRING":
+        format += `[$${token.value}]`;
+        break;
+      case "CHAR":
+        format += shouldEscapeFormatChar(token.value) ? `\\${token.value}` : token.value;
+        break;
+      default:
+        format += token.value;
+    }
+  }
+  return format;
+}
+
+function numberInternalFormatToTokenList(internalFormat: NumberInternalFormat): FormatToken[] {
+  let tokens: FormatToken[] = [...internalFormat.integerPart];
+
+  if (internalFormat.thousandsSeparator) {
+    const index = tokens.findIndex(
+      (token, i) => token.type === "DIGIT" && tokens[i + 1]?.type === "DIGIT"
```

Comment:
Yeah I'm aware of that, we don't always get the exact same format when parsing the format to an internal format => convert it back to a string (but functionnaly it should always do the same thing). I figured out that we don't really care.

But I guess I could leave the `THOUSAND_SEPARATORS` in the tokens and ignore them when applying the format. That also simplify a lot `numberInternalFormatToTokenList` so I guess it's not bad.

Reply:
It makes `_createLargeNumberFormat` more annoying tho, since we have to put the correct `THOUSAND_SEPARATORS` in the tokens, and cannot just play with internalNumberFormat.magnitude

Reply:
> I figured out that we don't really care.

100% true